corva/application.py

```diff
@@ -0,0 +1,52 @@
+from typing import Callable, List, Optional
+
+from corva.middleware.unpack_context import unpack_context
+from corva.models.base import BaseContext
+from corva.types import MIDDLEWARE_CALL_TYPE, MIDDLEWARE_TYPE
+
+
+def wrap_call_in_middleware(
+     call: Callable,
+     middleware: Optional[List[MIDDLEWARE_TYPE]] = None
+) -> MIDDLEWARE_CALL_TYPE:
+    def wrapper_factory(mw, call):
+        def wrapper(ctx):
+            return mw(ctx, call)
+
+        return wrapper
+
+    middleware = middleware or []
+
+    for mw in reversed(middleware):
+        call = wrapper_factory(mw, call)
+
+    return call
+
+
+class Corva:
+    def __init__(
+         self,
+         *,
+         middleware: Optional[List[MIDDLEWARE_TYPE[BaseContext]]] = None
+    ):
+        self.user_middleware = middleware or []
+
+    def get_middleware_stack(
+         self,
+         middleware: Optional[List[MIDDLEWARE_TYPE[BaseContext]]] = None
+    ) -> List[MIDDLEWARE_TYPE[BaseContext]]:
+        middleware = middleware or []
+
+        middleware_stack = (
+             middleware
+             + self.user_middleware
+             + [unpack_context]
```

Comment:
what is `unpack_context`? it makes no sense in the scope of this method and middleware stack

Reply:
it is a default middleware, that should always be called last. it scans parameter annotations of user function and puts values from context into argument names which have needed annotations. [link](https://github.com/corva-ai/python-sdk/blob/7e8a5aa1ceaf37e7447693b05cbc8f6c4ae0459a/docs_src/tutorial_1_hello_world.py#L7)

Reply:
inspired by [this](https://github.com/tiangolo/fastapi/blob/master/fastapi/dependencies/utils.py#L336)

Reply:
let's call it `default_middlewares` then?

Reply:
maybe adding comment will be enough, as `default_middleware` may be either before or after `user_middleware`. e.g [link](https://github.com/encode/starlette/blob/e4307065ea6dbe708fba9643d14fe7adfff06d46/starlette/applications.py#L79).

```suggestion
             + [unpack_context]  # default middleware, should always be called last
```

Reply:
why comment if we can rename the variable?

Reply:
it is not obvious whether `default_middleware` should be placed before of after `user_middleware`, especially, since there is another `middleware` in function. we can call it `tail_middleware` though

Reply:
let's call it `default_middleware` and write a comment `user-defined middlewares should come last.`